petasus/protein_ngram.py

```diff
@@ -0,0 +1,183 @@
+from __future__ import annotations
+
+import re
+from collections import defaultdict
+from os import PathLike
+
+from loguru import logger
+from pyteomics.fasta import FASTA
+from tqdm.auto import tqdm
+
+FASTA_NAME_REGEX = re.compile(r"^.*\|(.*)\|.*$")
+UNIPROT_ACC_REGEX = re.compile(r"^[A-Z0-9]{6}(-\d+)?$")
+
+
+class ProteinNGram:
+    """Implements an n-gram to fast lookup of proteins that match a peptide.
+
+    Usage
+    -----
+    ```python
+    ngram = ProteinNGram.from_fasta("path/to/fasta")
+    ngram.search_ngram("AAC")
+    ['Prot1']
+    ```
+
+    Examples
+    --------
+    >>> base_ngram = {'AA': {1,3}, 'AB': {2,3}, 'AC': {1, 4}, 'CA': {4}}
+    >>> inv_index = {1: "Prot1", 2: "Prot2", 3: "Prot3", 4: "Prot4"}
+    >>> inv_seq = {1: "AACAA", 2: "AABAA", 3: "ABDAA", 4: "CACAA"}
+    >>> ngram = ProteinNGram(
+    ...     ngram = base_ngram,
+    ...     inv_alias = inv_index,
+    ...     inv_seq = inv_seq)
+    >>> ngram.search_ngram("AAC")
+    ['Prot1']
+    >>> ngram.search_ngram("CAC")
+    ['Prot4']
+    """
+
+    __slots__ = ("ngram_size", "ngram", "inv_alias", "inv_seq")
+
+    def __init__(
+        self,
+        ngram: dict[str, set[int]],
+        inv_alias: dict[int, str],
+        inv_seq: dict[int, str],
+    ) -> None:
+        """Initialized an ngram for fast lookup.
+
+        For details check the main class docstring.
+        """
+        keys = list(ngram)
+        if not all(len(keys[0]) == len(k) for k in ngram):
+            raise ValueError("All ngram keys need to be the same length")
+        self.ngram_size: int = len(keys[0])
+        self.ngram = ngram
+        self.inv_alias = inv_alias
+        self.inv_seq = inv_seq
+
+    def search_ngram(self, entry: str) -> list[str]:
+        """Searches a sequence using the n-gram and returns the matches."""
+
+        if len(entry) < self.ngram_size:
+            raise ValueError(
+                f"Entry {entry} is shorter than the n-gram size ({self.ngram_size})"
+            )
+
+        candidates = None
+        for x in [
+            entry[x : x + self.ngram_size]
+            for x in range(1 + len(entry) - self.ngram_size)
+        ]:
+            if len(x) < self.ngram_size:
+                break
+
+            if candidates is None:
```

Comment:
This only works becuse the fasta builder uses a default dict instead of a dict, this can lead to key errors if not